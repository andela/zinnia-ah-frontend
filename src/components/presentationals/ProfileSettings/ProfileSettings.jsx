import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Form } from 'semantic-ui-react';
import { connect } from 'react-redux';
import { toast } from 'react-toastify';

import Button from '../../presentationals/Button/Button.jsx';
import { generateFormData, uploadImageToServer } from '../../../utils/helpers';
import {
  updateUserProfileRequest,
  getUserProfileRequest,
} from '../../../store/modules/profile';
import Loader from '../../presentationals/Loader/Loader.jsx';

class ProfileSettings extends Component {
  state = {
    isLoading: false,
    tabName: 'userSettings',
    previewImage: null,
  };

  componentDidMount() {
    this.props.currentView();
  }

  componentWillUnmount() {
    this.props.currentView('default');
  }

  componentDidUpdate = async prevProps => {
    if (this.props.data.updatedAt !== prevProps.data.updatedAt) {
      this.setState({ isLoading: false });
      await this.props.getUserProfileRequest(this.props.data.username);
    }
  };

  handleSubmit = event => {
    event.preventDefault();
    const formData = generateFormData(event.target);
    this.setState({ isLoading: true });
    if (this.state.previewImage) {
      return this.props.uploadImageToServer(
        {
          image: this.state.previewImage,
          tag: 'profile_pix',
        },
        (error, response) => {
          if (error) return toast.error(response, 'ERROR');
          formData.image = response;
          return this.props.updateUserProfileRequest(formData);
        },
      );
    }
    return this.props.updateUserProfileRequest(formData);
  };

  previewImage = ({ target }) => {
    const image = target.files[0];
    if (!image.type.startsWith('image/')) {
      return false;
    }
    const previewer = document.querySelector('.avatar-div .avatar.avatar-xl');

    const reader = new FileReader();
    reader.addEventListener('load', () => {
      previewer.src = reader.result;
    });
    reader.readAsDataURL(image);
    toast.success('Photo Ready to be saved to profile', 'DONE');
    this.setState({ previewImage: image });
  };

  render() {
    const { firstName, lastName, bio, username, email } = this.props.data;
    return (
      <div className="settings">
        <div className="edit-profile">
          {this.state.isLoading && (
            <Loader text="Updating Profile" size="large teal" />
          )}
          <Form onSubmit={this.handleSubmit}>
            <Form.Group widths="equal">
              <Form.Input
                fluid
                label="First name"
                placeholder="Jane"
                defaultValue={firstName}
                name="firstName"
              />
              <Form.Input
                fluid
                label="Last name"
                placeholder="Doe"
                defaultValue={lastName}
                name="lastName"
              />
            </Form.Group>
            <Form.Group widths="equal">
              <Form.Input
                fluid
                label="Email Address"
                placeholder="example@mail.com"
                defaultValue={email}
                type="email"
                name="email"
              />
              <Form.Input
                fluid
                label="Username"
                placeholder="janedoe1"
                defaultValue={username}
                name="username"
              />
            </Form.Group>
            <Form.TextArea
              label="Bio"
              name="bio"
              placeholder="Tell us more about you...(450 characters max)"
              maxLength="450"
              defaultValue={bio}
              rows={5}
            />
            <div
              className="field"
              style={{
                display: 'flex',
                justifyContent: 'flex-end',
              }}
            >
              <Button type="submit" value="SAVE CHANGES" className="btn-dark" />
              <Form.Input
                fluid
                type="file"
                accept="image/*"
                name="upload_profile_picture"
                placeholder="Select Profile Pix"
                onChange={this.previewImage}
              />
            </div>
          </Form>
        </div>
      </div>
    );
  }
}

ProfileSettings.propTypes = {
  currentView: PropTypes.func,
  uploadImageToServer: PropTypes.func.isRequired,
  updateUserProfileRequest: PropTypes.func.isRequired,
  getUserProfileRequest: PropTypes.func.isRequired,
  data: PropTypes.shape({
    firstName: PropTypes.string,
    lastName: PropTypes.string,
    username: PropTypes.string,
    email: PropTypes.string,
    updatedAt: PropTypes.string,
    bio: PropTypes.string,
  }),
};

const mapStateToProps = state => {
  return { data: state.profile };
};

export default connect(
  mapStateToProps,
  { updateUserProfileRequest, uploadImageToServer, getUserProfileRequest },
)(ProfileSettings);
