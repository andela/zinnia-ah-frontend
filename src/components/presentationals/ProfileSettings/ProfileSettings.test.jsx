import React from 'react';
import { mount } from 'enzyme';

import ProfileSettings from './ProfileSettings';

describe('Settings', () => {
  it('should render correctly', () => {
    const component = mount(<ProfileSettings />);
    expect(component).toMatchSnapshot();
  });
});
