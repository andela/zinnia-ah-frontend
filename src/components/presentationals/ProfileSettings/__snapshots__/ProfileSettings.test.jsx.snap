// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileSettings should render profile settings correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ConnectFunction
    currentView={[Function]}
  >
    <ProfileSettings
      currentView={[Function]}
      data={
        Object {
          "profile": Object {
            "currentView": [Function],
            "data": Object {
              "email": "example@mail.com",
              "username": "examplename",
            },
            "message": "fetched user",
            "status": "success",
          },
        }
      }
      getUserProfileRequest={[Function]}
      updateUserProfileRequest={[Function]}
      uploadImageToServer={[Function]}
    >
      <div
        className="settings"
      >
        <div
          className="edit-profile"
        >
          <Form
            as="form"
            onSubmit={[Function]}
          >
            <form
              className="ui form"
              onSubmit={[Function]}
            >
              <FormGroup
                widths="equal"
              >
                <div
                  className="equal width fields"
                >
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    fluid={true}
                    label="First name"
                    name="firstName"
                    placeholder="Jane"
                  >
                    <FormField
                      control={[Function]}
                      fluid={true}
                      label="First name"
                      name="firstName"
                      placeholder="Jane"
                    >
                      <div
                        className="field"
                      >
                        <label>
                          First name
                        </label>
                        <Input
                          fluid={true}
                          name="firstName"
                          placeholder="Jane"
                          type="text"
                        >
                          <div
                            className="ui fluid input"
                          >
                            <input
                              name="firstName"
                              onChange={[Function]}
                              placeholder="Jane"
                              type="text"
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                  </FormInput>
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    fluid={true}
                    label="Last name"
                    name="lastName"
                    placeholder="Doe"
                  >
                    <FormField
                      control={[Function]}
                      fluid={true}
                      label="Last name"
                      name="lastName"
                      placeholder="Doe"
                    >
                      <div
                        className="field"
                      >
                        <label>
                          Last name
                        </label>
                        <Input
                          fluid={true}
                          name="lastName"
                          placeholder="Doe"
                          type="text"
                        >
                          <div
                            className="ui fluid input"
                          >
                            <input
                              name="lastName"
                              onChange={[Function]}
                              placeholder="Doe"
                              type="text"
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                  </FormInput>
                </div>
              </FormGroup>
              <FormGroup
                widths="equal"
              >
                <div
                  className="equal width fields"
                >
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    fluid={true}
                    label="Email Address"
                    name="email"
                    placeholder="example@mail.com"
                    type="email"
                  >
                    <FormField
                      control={[Function]}
                      fluid={true}
                      label="Email Address"
                      name="email"
                      placeholder="example@mail.com"
                      type="email"
                    >
                      <div
                        className="field"
                      >
                        <label>
                          Email Address
                        </label>
                        <Input
                          fluid={true}
                          name="email"
                          placeholder="example@mail.com"
                          type="email"
                        >
                          <div
                            className="ui fluid input"
                          >
                            <input
                              name="email"
                              onChange={[Function]}
                              placeholder="example@mail.com"
                              type="email"
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                  </FormInput>
                  <FormInput
                    as={[Function]}
                    control={[Function]}
                    fluid={true}
                    label="Username"
                    name="username"
                    placeholder="janedoe1"
                  >
                    <FormField
                      control={[Function]}
                      fluid={true}
                      label="Username"
                      name="username"
                      placeholder="janedoe1"
                    >
                      <div
                        className="field"
                      >
                        <label>
                          Username
                        </label>
                        <Input
                          fluid={true}
                          name="username"
                          placeholder="janedoe1"
                          type="text"
                        >
                          <div
                            className="ui fluid input"
                          >
                            <input
                              name="username"
                              onChange={[Function]}
                              placeholder="janedoe1"
                              type="text"
                            />
                          </div>
                        </Input>
                      </div>
                    </FormField>
                  </FormInput>
                </div>
              </FormGroup>
              <FormTextArea
                as={[Function]}
                control={[Function]}
                label="Bio"
                maxLength="450"
                name="bio"
                placeholder="Tell us more about you...(450 characters max)"
                rows={5}
              >
                <FormField
                  control={[Function]}
                  label="Bio"
                  maxLength="450"
                  name="bio"
                  placeholder="Tell us more about you...(450 characters max)"
                  rows={5}
                >
                  <div
                    className="field"
                  >
                    <label>
                      Bio
                    </label>
                    <TextArea
                      as="textarea"
                      maxLength="450"
                      name="bio"
                      placeholder="Tell us more about you...(450 characters max)"
                      rows={5}
                    >
                      <Ref
                        innerRef={
                          Object {
                            "current": <textarea
                              maxlength="450"
                              name="bio"
                              placeholder="Tell us more about you...(450 characters max)"
                              rows="5"
                            />,
                          }
                        }
                      >
                        <RefFindNode
                          innerRef={
                            Object {
                              "current": <textarea
                                maxlength="450"
                                name="bio"
                                placeholder="Tell us more about you...(450 characters max)"
                                rows="5"
                              />,
                            }
                          }
                        >
                          <textarea
                            maxLength="450"
                            name="bio"
                            onChange={[Function]}
                            onInput={[Function]}
                            placeholder="Tell us more about you...(450 characters max)"
                            rows={5}
                          />
                        </RefFindNode>
                      </Ref>
                    </TextArea>
                  </div>
                </FormField>
              </FormTextArea>
              <div
                className="field"
                style={
                  Object {
                    "display": "flex",
                    "justifyContent": "flex-end",
                  }
                }
              >
                <Button
                  className="btn-dark"
                  type="submit"
                  value="SAVE CHANGES"
                >
                  <button
                    className="btn btn-dark"
                    type="submit"
                  >
                    SAVE CHANGES
                  </button>
                </Button>
                <FormInput
                  accept="image/*"
                  as={[Function]}
                  control={[Function]}
                  fluid={true}
                  name="upload_profile_picture"
                  onChange={[Function]}
                  placeholder="Select Profile Pix"
                  type="file"
                >
                  <FormField
                    accept="image/*"
                    control={[Function]}
                    fluid={true}
                    name="upload_profile_picture"
                    onChange={[Function]}
                    placeholder="Select Profile Pix"
                    type="file"
                  >
                    <div
                      className="field"
                    >
                      <Input
                        accept="image/*"
                        fluid={true}
                        name="upload_profile_picture"
                        onChange={[Function]}
                        placeholder="Select Profile Pix"
                        type="file"
                      >
                        <div
                          className="ui fluid input"
                        >
                          <input
                            accept="image/*"
                            name="upload_profile_picture"
                            onChange={[Function]}
                            placeholder="Select Profile Pix"
                            type="file"
                          />
                        </div>
                      </Input>
                    </div>
                  </FormField>
                </FormInput>
              </div>
            </form>
          </Form>
        </div>
      </div>
    </ProfileSettings>
  </ConnectFunction>
</Provider>
`;
